File: analyze/train.py
import csv
import subprocess
from subprocess import PIPE


file_path = "./results.csv"
train_num = 2

def build():
  subprocess.run("generator.exe test 20 1 8", stderr=PIPE)
  with open("in.txt", "w") as f:
    subprocess.run("tester.exe test", stdout=f)

def run_program():
  build()
  subprocess.run("g++ yakinamashi_thread.cpp -Ofast")
  with open("in.txt", "r") as f:
    process = subprocess.run("a.exe", stdin=f, stdout=PIPE, stderr=PIPE, text=True)
  result = process.stdout
  print(result, end='')
  score, audio, karuta = map(int, result.split())
  return score, audio, karuta

def main():
  with open(file_path, "a", newline='\n') as write_file:
    writer = csv.writer(write_file)
    for i in range(train_num):
      print("start:", i)
      score, audio, karuta = run_program()
      writer.writerow([ score, audio, karuta ])
  

if __name__ == "__main__":
  subprocess.run("g++ generator.cpp -O2 -o generator.exe")
  subprocess.run("g++ tester.cpp -O2 -o tester.exe")
  main()

File: analyze/train.cpp
#include <iostream>
#include <string>
#include <fstream>

void init(){
#if defined(_WIN32) || defined(_WIN64)
  system("g++ generator.cpp -O2 -o generator.exe");
  system("g++ run.cpp -O2 -o run.exe");
#else
  system("g++ generator.cpp -O2 -o generator");
  system("g++ run.cpp -O2 -o run");
#endif
}

void run_program(const int seed = -1){
  char buf[64];
#if defined(_WIN32) || defined(_WIN64)
  if(seed == -1) system("generator.exe test 20 1 8");
  else{
    sprintf(buf, "generator.exe test 20 1 8 %d", seed);
    system(buf);
  }
  system("run.exe > res.txt");
#else
  if(seed == -1) system("./generator test 20 1 8");
  else{
    sprintf(buf, "./generator test 20 1 8 %d", seed);
    system(buf);
  }
  system("./run > res.txt");
#endif
}

int main(int argc, char *argv[]){
  if(argc != 3){
    printf("error: train.exe [file path] [train num]");
    return 0;
  }
  const int train_num = std::stoi(argv[2]);
  init();
  for(int i = 0; i < train_num; i++){
    printf("START: %d\n", i+1);
    run_program(i);
    std::ofstream result(argv[1], std::ios::app);
    std::ifstream output("res.txt");
    int audio_diff, karuta_diff, score;
    output >> audio_diff >> karuta_diff >> score;
    result << audio_diff << "," << karuta_diff << "," << score << "\n";
    output.close();
    result.close();
  }
}